### PR TITLE
FUTURE: initial support for .npmrc file

### PR DESCRIPTION
**Pull request by bartlomieju**
This commit adds initial support for ".npmrc" files.

Currently we only discover ".npmrc" files next to "package.json" files
and discovering these files in user home dir is left for a follow up.

This pass supports "_authToken" and "_auth" configuration
for providing authentication.

LSP support has been left for a follow up PR.

Towards https://github.com/denoland/deno/issues/16105

### CODE REVIEW

File: cli/http_util.rs [hunk 1]
```diff
@@ -339,15 +345,24 @@ impl HttpClient {
...
     if status.is_redirection() {
       for _ in 0..5 {
         let new_url = resolve_redirect_from_response(&url, &response)?;
-        let new_response =
-          self.get_no_redirect(new_url.clone())?.send().await?;
+        let mut builder = self.get_no_redirect(new_url.clone())?;
+        if let Some((header_name, header_value)) = maybe_header.as_ref() {
+          builder = builder.header(header_name, header_value);
+        }
+        let new_response = builder.send().await?;
```
Comment (lucacasonato) [inline cli/http_util.rs:365]: You should only serve auth tokens after a redirect if the original and redirected URL have the same origin.

File: cli/npm/cache_dir.rs [hunk 2]
```diff
@@ -97,23 +104,37 @@ impl NpmCacheDir {
...
-        "{}/",
-        root_url_to_safe_local_dirname(registry_url)
-          .to_string_lossy()
-          .replace('\\', "/")
-      ))
-      // this not succeeding indicates a fatal issue, so unwrap
-      .unwrap();
-    let mut relative_url = registry_root_dir.make_relative(specifier)?;
-    if relative_url.starts_with("../") {
-      return None;
```
Comment (dsherret) [inline cli/npm/cache_dir.rs:114]: I think we could make this more optimal by storing `root_url_to_safe_local_dirname` for each registry url in the constructor, then here getting the relative path from the `self.root_dir_url`, then check the first component for a match, then continue below

File: cli/npm/managed/resolvers/mod.rs [hunk 3]
```diff
@@ -25,7 +24,6 @@ pub fn create_npm_fs_resolver(
   fs: Arc<dyn FileSystem>,
   cache: Arc<NpmCache>,
   progress_bar: &ProgressBar,
-  registry_url: Url,
```
Comment (dsherret) [inline cli/npm/managed/resolvers/mod.rs:28]: It's nice that these are being removed.

File: cli/standalone/mod.rs [hunk 4]
```diff
@@ -447,8 +450,9 @@ pub async fn run(
               CliNpmResolverManagedPackageJsonInstallerOption::ConditionalInstall(
                 package_json_deps_provider.clone(),
               ),
-            npm_registry_url,
             npm_system_info: Default::default(),
+            // TODO(bartlomieju): do we need to support it here?
+            npmrc: create_default_npmrc(),
```
Comment (dsherret) [inline cli/standalone/mod.rs:455]: I don't think we need it here. The package information stored should be resolved based on the user's npmrc locally, but in the compiled executable it should know nothing about that. I think you can remove the todo and maybe add a comment that says the packages from different registries are already inlined or something.

File: cli/standalone/mod.rs
Comment (dsherret) [inline cli/standalone/mod.rs:1]: It would be good to add a deno compile test. I think it should just work now?

**Review by dsherret — COMMENTED** [review]
Nice! Looks really good. Just a few comments.

File: cli/standalone/mod.rs
Comment (bartlomieju) [reply cli/standalone/mod.rs:1]: It appears it just works 👍

**Review by bartlomieju — COMMENTED** [review]
It appears `--reload` flag is not working correctly if `.npmrc` is present and custom registries are used.

This is a blocker to landing this PR.

**Review by dsherret — APPROVED** [review]
LGTM!

**Pull request merged**